packages/core/src/task.ts

```diff
@@ -26,180 +22,221 @@ export interface TaskOptions {
 
 type EnsureHandler = () => void;
 
-export class Task<TOut = unknown> extends EventEmitter implements Promise<TOut>, Trapper {
-  public id = ++COUNTER;
+type WithControls<TOut> = { [CONTROLS]?: Controls<TOut> }
+
+export interface Task<TOut = unknown> extends Promise<TOut> {
+  readonly id: number;
+  readonly state: State;
+  readonly result?: TOut;
+  readonly error?: Error;
+  catchHalt(): Promise<TOut | undefined>;
```

Comment:
I've never needed to use this in public code, but feels like if we are going to make it public, then the return type should be`Promise<MaybeHalted<T>>` because if this is a `Task<undefined>` then there is no way to distinguish between the halt and resolve case. E.g.

```ts
type MaybeHalted<T> = { didHalt: true, value: T } | { didHalt: false, value: undefined };

interface Task<T> {
  /*..../*
  maybe: Promise<MaybeHalted<T>>;
```

Does this need to be part of the public API for now?

Reply:
I personally find this method really useful. The point of it isn't so much to provide a useful return value, and more for halts to not blow anything up. Since `HaltError` is not exported, there isn't any other really safe way of preventing this.

Reply:
I can definitely see why it's useful, especially when you're consuming a task from outside effection. But what about the case though where you have a `Task<undefined>`. Is it useful to disambiguate between `undefined` because successful completion, or `undefined` because of halt?